src/overview/overview.md

```diff
@@ -0,0 +1,643 @@
+- Method calls are desugared into regular function calls.
+- Overloaded operators like `+` are lowered into function calls or builtin
+  operations.
+- Match ergonomics are desugared.
+
+With the aforementioned properties, the construction of THIR synthesizes enough
+information for MIR building, and makes lowering to MIR easier and less
+error-prone than if HIR was attempted to be directly lowered to MIR.
+
+See [The Typed High-level Intermediate
+Representation](https://rustc-dev-guide.rust-lang.org/thir.html) for more
+details.
+
+#### HIR Body Type-checking
+
+> 🚧 **TODO** 🚧 (Remove the cat meme when complete with draft)
+
+![](https://hackmd.io/_uploads/ry2CM6gAa.jpg)
+
+High-level Intermediate Representation (HIR) is a higher-level IR than THIR.
+When compared to MIR, HIR resembles the surface Rust syntax more closely. When
+compared to AST, HIR has desugared constructs (such as desugaring `for` loops
+into `loop`s) to make analysis easier.
+
+The goal of HIR body typechecking is to infer types and other information
+necessary for THIR/MIR building. HIR body type-checking recursively walks and
+checks each expression. This process produces information such as the type of
+each expression. During the process, type inference, implicit cocercions, trait
+bound checking and method resolution is performed. Notably,
+
+- Method resolution is not performed in (early/late) Name Resolution because
+  type information is required.
+- Type inference is performed on HIR and not MIR because HIR maps closer to the
+  surface Rust syntax (and retains the tree-based structure), which enables more
+  descriptive errors and suggestions to be produced.
+- MIR building requires method resolution information from HIR type-checking to
+  guide the lowering.
+
+When compared to MIR typeck where all types are specified, HIR typeck has to
+infer types. HIR typeck also checks all closures together with their parent body
+while MIR typeck/borrowck checks them separately.
+
+See [Type checking](https://rustc-dev-guide.rust-lang.org/type-checking.html)
+for more details.
+
+#### HIR Well-formedness Checking
+
+A type is _well-formed_ (WF) when it satisfies a set of requirements which are
+fundamental to the type's validity. The requirements are type-specific, for
+example:
+
+- For a tuple to be WF, all elements except for the last needs to be `Sized`.
+- For an algebraic data type (struct/enum/union) to be WF, its generics need to
+  satisfy the `where` clauses on the ADT.
+
+The notion of well-formedness can be inductively extended for entities that
+contain types, such as trait predicates. A trait predicate `T0: Trait<P1, ...>`
+is well-formed if `T0` and `P1, ...` are well-formed in addition to `Trait`'s
+`where` clauses being satisfied.
+
+HIR well-formedness checking is an early collection of checks on mostly surface
+Rust syntax (such as parameters, local variables and return types) which are run
+before further type-checking and inference, for two purposes:
```

Comment:
```suggestion
The HIR is _well-formed_ (WF) when it satisfies a set of requirements that include:
- Well-formedness of types in type signatures
- Object safety
- Function call ABI requirements (e.g. for "rust-call")
- Check if receiver type can be used as a self type
- Variances check
- etc
These checks are run before further type-checking and inference, for two purposes:
```
(I just took the list from your Zulip post, I don't know if that's a sensible list)

Reply:
I added these examples to extend the notion of HIR wfcheck to cover not just type wfcheck.